Stats Widget: Users without a site are incorrectly asked to login by the widget

staskus opened the pull request:
## Issue

Closes #15857 

## Description

Users without a site are asked to login by the Stats Widget

### Rootcause

This case is not explicitly handled in the Stats Widget.

### Solution

1. Update `SiteListProvider` `getTimelines` method to reflect the logic:
- If user doesn't have default site and is logged in -> "Create or add a site to see today's/this week's/all time stats."
- If user doesn't have default site and is not logged in -> "Log in to WordPress to see today's stats"

2. Set `WPStatsHomeWidgetsUserDefaultsSiteIdKey` to `defaultSiteId` and `nil` in `StatsWidgetStore`. Since this key is only designed to be used by `StatsWidget` it makes sense to set it in `StatsWidgetStore` so full solution would be contained within this class and parts not scattered in `AccountService`.

3. Add `WPSiteDeleted` and `WPSiteCreated` notifications that are posted from services and listened in `StatsWidgetStore`. It is done for 2 main reasons:
- Avoid `Services` being aware and depend on `StatsWidgetStore`. They are lower level abstractions and should not care about the `Widgets`
- Increase understanding of `StatsWidgetStore` functionality. We can clearly see on which state changes the `StatsWidget` depends on and how we react.
 
### Another issue

While working on this issue found another issue with no sites case and widgets: [Stats Widget does not load when a new site appears after refreshing "My Site" tab](https://github.com/wordpress-mobile/WordPress-iOS/issues/18914) . It's more of a corner case and could potentially cause multiple changes in the codebase so decided not to tackle it in this PR.

## Testing instructions

#### 1st scenario - Add Widget -> Log in to an account without any site

1. Create WordPress account without any site
3. Make sure you are logged out of the app
4. Add WordPress "Stats Widget" (All Time / This Week / Today) on the home screen
5. Confirm that "Log in to WordPress..." message is shown on the widget
6. Log in into the app
7. Go back to the home screen
8. Confirm that "Create or add a site to see today's/this week's/all time stats."" message is shown on the widget

#### 2nd scenario -  Log in to an account without any site -> Add Widget
1. Make sure you are logged in into account without any site
2. Add WordPress "Stats Widget" (All Time / This Week / Today) on the home screen
3. Confirm that "Create or add a site to see today's/this week's/all time stats." message is shown on the widget

#### 3rd scenario - Log in to an account without any site -> Add Widget -> Create a new site
_As an external contributor I wasn't able to test this scenario due to Public-API limitations_

1. Make sure you are logged in into account without any site
2. Add WordPress "Stats Widget" (All Time / This Week / Today) on the home screen
3. Create a new site on the app
4. Go back to the home screen
5. Confirm that Stats are shown on the widget

#### 4th scenario - Log in to an account without any site -> Add Widget -> Log out:
1. Make sure you are logged in into account without any site
2. Add WordPress "Stats Widget" (All Time / This Week / Today) on the home screen
3. Confirm that "Create or add a site to see today's/this week's/all time stats." message is shown on the widget
4. Log out from the app
5. Confirm that "Log in to WordPress..." message is shown on the widget6. 

#### 5th scenario - Log in into account that has a site -> Remove site

1. Log in into an account that does have a site.
2. Add Stats widgets.
3. Delete the site (via My Site (menu tab selected) > Site Settings > Delete Site).
4. Confirm that "Create or add a site to see today's/this week's/all time stats." message is shown on the widget.

## Regression Notes

1. Potential unintended areas of impact

The other states of the widget

2. What I did to test those areas of impact (or what existing automated tests I relied on)

Retested scenarios in https://github.com/wordpress-mobile/WordPress-iOS/pull/18885

3. What automated tests I added (or what prevented me from doing so)

It may make sense to test `SiteListProviderTests` `getTimelines` as it has a complicated business logic. However, extension unit-testing has limitations and a recommended way would be to have a framework that is used by an extension and test that code instead. 

## Video & Photos

<img width="553" alt="image" src="https://user-images.githubusercontent.com/4062343/176404631-f9d6a2e4-8877-43e4-9990-71e1373a0ba0.png">

https://user-images.githubusercontent.com/4062343/176403517-d7c6d7ae-d0a8-4b59-b92d-cc33b6a5402e.mov

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

staskus committed:
Remove defaultSiteID from UserDefaults during logout

staskus committed:
Added localizable string for no site state in the widget

The description mentions both options of creating and adding an existing site to see the stats.

staskus committed:
Added noSite state for StatsWidgetView

staskus committed:
No site message in Stats widget if user is logged in without a default site

staskus commented:
> I have considered if this change warrants user-facing release notes and have added them to RELEASE-NOTES.txt if necessary.

After looking at the`RELEASE-NOTES.txt` I very subjectively decided that this is too small of a change to "warant" user-facing notes.

I also see that we have `[internal]` flag in the `RELEASE-NOTES.txt` are these meant not to be seen by the general public?  If so, what "warrants" adding them there?

guarani commented on WordPress/WordPressStatsWidgets/Views/Localization/LocalizableStrings.swift:
```diff
@@ -57,6 +57,11 @@ enum LocalizableStrings {
 
+    // No site view
+    static let noSiteViewTitle = AppLocalizedString("widget.today.nosite.view.title",
+                                                    value: "Create or add a site to see the stats.",
```
```suggestion
                                                    value: "Create or add a site to see stats.",
```

staskus replied:
Thanks. 

Regarding message: Yes, I think it makes sense to improve localization to have it be more consistent and include widget type. 

Regarding site creation: Thanks for testing it. It looks like notifications are not posted during the site creation. I'll think how I could better tackle it.

guarani commented:
> #### 4th scenario - Log in to an account without any site -> Add Widget -> Log out:
> 
>     1. Make sure you are logged in into account without any site
> 
>     2. Add WordPress "Stats Widget" (All Time / This Week / Today) on the home screen
> 
>     3. Confirm that "Log in to WordPress..." message is shown on the widget

At step 3 of the 4th scenario here, should the message be "Create or add a site to see the stats." instead (given the user is logged in without a site)?

staskus commented:
> > #### 4th scenario - Log in to an account without any site -> Add Widget -> Log out:
> > ```
> > 1. Make sure you are logged in into account without any site
> > 
> > 2. Add WordPress "Stats Widget" (All Time / This Week / Today) on the home screen
> > 
> > 3. Confirm that "Log in to WordPress..." message is shown on the widget
> > ```
> 
> At step 3 of the 4th scenario here, should the message be "Create or add a site to see the stats." instead (given the user is logged in without a site)?

@guarani yes, you are right. I'll update the testing scenarios.

staskus committed:
Merge remote-tracking branch 'upstream/trunk' into issue/15857-users-without-a-site-are-incorrectly-asked-to-login-by-the-widget

staskus committed:
Set WPStatsHomeWidgetsUserDefaultsSiteIdKey in StatsWidgetStore

Since WPStatsHomeWidgetsUserDefaultsSiteIdKey key is only used in the Stats Widget context it makes more sense to tie it to related Stats Widget events and not to bury it in more generic AccountService class.

We also have expectation for site data to be loaded after logInNotification and site data to be deleted after defaultSiteChanged notification, so setting user defaults key and then using it afterwards brings more transparency on what Stats widget depends on.

staskus commented:
1. Merged with the other Stats widget changes and retested the solution
2. Set and removed `WPStatsHomeWidgetsUserDefaultsSiteIdKey` in `StatsWidgetStore` instead of `AccountService`. Since this key is only designed to be used by `StatsWidget` it makes sense to set and remove it in `StatsWidgetStore` so full solution would be contained within this class and parts not scattered in `AccountService`.

CC: @guarani

guarani reviewed (changes requested):
Hi @staskus, thanks for the PR! 

I triggered CI via a separate PR, so if that passes, the checks on this PR should also turn green. 

### Inconsistency in messaging

| "No site" scenario | "Not logged in" scenario |
| - | - |
| <img src="https://user-images.githubusercontent.com/1898325/176214988-9ce9de88-cb2c-49b1-b032-46fd59af6667.jpeg" alt="" width="350">  | <img src="https://user-images.githubusercontent.com/1898325/176214949-6d9f3ff6-19cf-46ac-a800-e967475f4fe8.jpeg" alt="" width="350">  |

In the "No site" scenario, could the message include the widget type (like it does in the "Not logged in scenario")? Also, with regard to the copy here, for clarity, I'd change "... see the stats" to just "... see stats".

In the scenario which requires site creation (which you mentioned you can't test due to an API limitation of your WordPress.com developer account), the widgets didn't update after I created a site. They kept showing "Create or add a site to see the stats.".
I edited the widget and "reselected" the site, but the widget didn't update to show stats. I then deleted and re-added the widget and it still didn't work.

I didn't test the remaining scenarios, given the above. While you may be able to address the first issue directly, please reach out on Slack regarding the second issue (and let me know if you have any ideas for working around it).

staskus committed:
Expanding noSite state for today, this week and all time

- Remove noSite and noStats for StatsWidgetKind. It should only contain today, thisWeek and allTime
- Passing StatsWidgetKind for noSite, same way as for loggedOut case

staskus committed:
Expanding noData state for today, this week and all time

staskus committed:
Set defaultSiteId for Stats widget when creating new site

- Set defaultSiteId in initializeStatsWidgetsSiteList method. It's safer in general to update defaultSiteId if any entity prompts to initialize or refresh statsWidgetSiteList
- SiteAssemblyService calls refreshStatsWidgetsSiteList after site creation

staskus commented:
1. Expanded noSite state to show text based on the widget type
2. Expanded noData state to show text based on the widget type (for the sake of consistency)
3. Fixed the issue when Stats widget is not configured after creating a new site. The problem was that `defaultSiteId` wasn't set as it didn't exist yet after `logIn` notification. `SiteAssemblyService` calls Stats widget initialization method after creating a site, so from StatsWidgetStore side it's important to make sure `defaultSiteId` is set when someone tries to initialize Stats widget. 
Tested on simulator by rewriting site creation request on Charles to use prod clientId & clientSecret.

https://user-images.githubusercontent.com/4062343/176403517-d7c6d7ae-d0a8-4b59-b92d-cc33b6a5402e.mov

guarani reviewed (changes requested):
Hi @staskus, 

> 3\. Fixed the issue when Stats widget is not configured after creating a new site. The problem was that `defaultSiteId` wasn't set as it didn't exist yet after `logIn` notification. `SiteAssemblyService` calls Stats widget initialization method after creating a site, so from StatsWidgetStore side it's important to make sure `defaultSiteId` is set when someone tries to initialize Stats widget.
>         Tested on simulator by rewriting site creation request on Charles to use prod clientId & clientSecret.

Nice job working around the site creation API limitation to fix this issue! 🙇 I retested scenario 3 and it's working. 

I noticed an issue in the following scenario: 
1. I logged into an account that _did_ have a site, and then deleted the site (via My Site (menu tab selected) > Site Settings > Delete Site)
2. I'm now still logged in, but have no sites
3. I add the any Stats widget
4. The widget shows stats for the recently deleted site, I expected it to display the "Create or add a site..." message.

Here's a screen recording:

https://user-images.githubusercontent.com/1898325/176567399-d4b9d3d1-c0c0-493e-9dc0-b8f329134081.mov

guarani commented on WordPress/WordPressStatsWidgets/Views/Localization/LocalizableStrings.swift:
```diff
@@ -57,6 +57,31 @@ enum LocalizableStrings {
 
+    static let noDataViewTodayTitle = AppLocalizedString("widget.today.nodata.view.title",
+                                                         value: "Unable to load today's stats.",
+                                                         comment: "Title of the nodata view in today widget")
```
nitpick: I think changing this to "...the 'no data' view..." will provide more clarity to translators.

staskus commented:
@guarani, thanks for testing the case when the site is deleted. I admit I didn't think about it. It a little bit reminds of https://github.com/wordpress-mobile/WordPress-iOS/issues/18914 . I'll explore at least 2 different ways to see which makes sense more:

1. A direct call to refreshWidget when the site is deleted (in the same way it's called when the site is created)
2. Listen in the StatsWidgetStore to some sort of notification that notifies when site is added / deleted so all the current cases and unforseen cases would be automatically handled.

staskus committed:
Refresh Stats widget after the site is removed

- Creating WPSiteDeleted notification
- Throwing notification from SiteManagementService

- Avoided polluting SiteManagementService with Widget code
- Ensure that Stats widget is refreshed if site is deleted from any place

staskus committed:
Remove StatsWidgetStore dependency from SiteAssemblyService

To keep the StatsWidgetStore solution consistent, throw WPSiteCreated notification instead of calling StatsWidgetStore directly.

We will be able to call this notification in other parts of application that creates a site or detects a new site being created.

staskus committed:
Update Stats Widget localization string comments

staskus committed:
Remove unnecessary spacing in SiteListProvider

staskus commented:
Thanks again for the comments. It's really an interesting problem how to better structure `StatsWidget` code to react to updates of the app state. 

The question is which state changes are relevant and which don't. We could also listen to `blogUpdated` notification. For example, if we change the name of the site and we want it to be immediately reflected on the widget. What I found is that `blogUpdated` notification might be posted many times throughout the app flow and constant calls to `reloadTimelines` then causes some meaningful updates of the widget then happen after delay. If we want to handle more cases I suggest addressing it on the different ticket so they could be properly reviewed. 

For now:
Added `WPSiteDeleted` and `WPSiteCreated` notifications that are posted from services and listened in `StatsWidgetStore`. It is done for 2 main reasons:
- Avoid `Services` being aware and depend on `StatsWidgetStore`. They are lower level abstractions and should not care about the `Widgets`
- Increase understanding of `StatsWidgetStore` functionality. We can clearly see on which state changes the `StatsWidget` depends on and how we react.

https://user-images.githubusercontent.com/4062343/176652439-577a1219-6055-4953-8867-70450a650107.mov

Such changes will possibly affect how https://github.com/wordpress-mobile/WordPress-iOS/issues/18914 is tackled, I updated the description. Also included "Site deletion" case as well to be checked.

guarani commented:
> I also see that we have `[internal]` flag in the `RELEASE-NOTES.txt` are these meant not to be seen by the general public? If so, what "warrants" adding them there?

Good question! Sorry I didn't notice this earlier. 

Using `[internal]` at the beginning of the description will indicate this should go in calls for beta testing, but not public release notes.

guarani reviewed (commented):
> 5th scenario - Log in into account that has a site -> Remove site

When I log into an account that has a site, and then I add a widget, the widget shows the message saying I don't have any sites:

https://user-images.githubusercontent.com/1898325/176756107-b3090066-18c5-4ed7-94d7-331dbb8b687b.mov

I expected it to show a populated widget. Can you reproduce this?

guarani commented:
The CI tests failed, but I believe it's due to a flaky test:

```
Failing tests:
--
		PostServiceWPComTests.testTrashingAPostWillUpdateItsRevisionStatusAfterSyncProperty()
```

I hit restart now ⏳

staskus commented:
> > 5th scenario - Log in into account that has a site -> Remove site
> 
> When I log into an account that has a site, and then I add a widget, the widget shows the message saying I don't have any sites:
> 
>  RPReplay_Final1656614453.mov 
> 
> I expected it to show a populated widget. Can you reproduce this?

Thanks for finding this issue, @guarani ! Yes, it should definetely show populated widget. I went through this scenario multiple times and I cannot reproduce it. I thought it might somehow be reproducible by logging in to different accounts, when the first account doesn't have a site and the next one does. I again used Charles to be able to log in into different accounts within the same session without killing the app but I wasn't able to reproduce it. 

Does the app always behave for you like this or only in some specific scenario or some specific account? Maybe it would give me clues.

By looking at the code:
- If `defaultSiteID` is set but `HomeWidgetData` is not loaded, we woud see `noData` view instead of `noSites` view.
- If both `defaultSiteID` and `HomeWidgetData` are set but widget is not reloaded after login, then we would see `log in` view rather than `no sites` view .
-  It could be like this if `WPStatsHomeWidgetsUserDefaultsSiteIdKey` is not set after log in. Meaning that `defaultSiteID` is `nil` when `WPSigninDidFinishNotification` is posted. For now, I cannot identify the scenario when this could happen.

staskus commented:
@guarani 

Do you know the setup you had or more details about the scenario to reproduce it?

> When I log into an account that has a site, and then I add a widget, the widget shows the message saying I don't have any sites: 

I didn't find a way to reproduce it. Thanks! 🙏

guarani commented:
Hi @staskus, when I ran into this, I created a new account + site. I don't recall anything special about the setup. Perhaps the Charles proxy is the key differentiator between my and your setups, but I'm not sure.

The temporary PR I created to trigger CI checks on your commits is over at https://github.com/wordpress-mobile/WordPress-iOS/pull/18942. I think site creation might work on the installable builds there, in case you want to give it a try there.

mokagio commented:
Pushing this PR to the next milestone, 20.4, because I'm running the 20.3 code-freeze today and this hasn't been approved yet.

guarani commented:
Thanks @mokagio! I also removed the milestone now because this isn't being worked on right now.

staskus commented:
Closing in favor of https://github.com/wordpress-mobile/WordPress-iOS/pull/19962 which is not merged from fork
